File: py-project/Solution59-2.py
# -*- coding:utf-8 -*-
class QueueMax:
    def __init__(self):
        self.data = []
        self.data_max = []

    def push_back(self,num):
        self.data.append(num)
        while self.data_max and self.data_max[-1] <= num:
            self.data_max.pop()
        self.data_max.append(num)
        return

    def pop_front(self):
        if not self.data:
            raise Exception('Empty Queue Cannot Pop')
        if self.data[0] == self.data_max[0]:
            self.data_max.pop(0)
        return self.data.pop(0)

    def max(self):
        return self.data_max[0]

    # def printData(self):
    #     return self.data
    #
    # def printDataMax(self):
    #     return self.data_max


if __name__ == '__main__':
    k = [2,3,4,2,6,2,5,1]
    n = 3
    # x = list(k)x
    # print(''.join(x))
    solution1 = QueueMax()
    solution1.push_back(2)
    print(solution1.data)
    print(solution1.data_max)
    print(solution1.max())
    solution1.push_back(4)
    print(solution1.data)
    print(solution1.data_max)
    print(solution1.max())
    solution1.push_back(3)
    print(solution1.data)
    print(solution1.data_max)
    print(solution1.max())
    solution1.push_back(1)
    print(solution1.data)
    print(solution1.data_max)
    print(solution1.max())
    solution1.pop_front()
    print(solution1.data)
    print(solution1.data_max)
    print(solution1.max())
    solution1.pop_front()
    print(solution1.data)
    print(solution1.data_max)
    print(solution1.max())
    solution1.pop_front()
    print(solution1.data)
    print(solution1.data_max)
    print(solution1.max())

File: c2-project/Solution59-2.cpp
//
// Created by 陈高杰 on 2020/2/20.
//
#include <iostream>
#include <stdlib.h>
#include <string>
#include <vector>
#include <stack>
#include <queue>
#include <math.h>
#include <sstream>
#include <map>
using namespace std;


// 送入大的值则吧之前比起小的都替换了
// 不然就加进去，因为之后可能会弹出最大值
class QueueMax {
public:
    deque<int> data;
    deque<int> data_max;
    int max(){
        return data_max[0];
    }
    void push_back(int num){
        data.push_back(num);
        while(!data_max.empty() && data_max[-1]<=num)
            data_max.pop_back();
        data_max.push_back(num);
        return;
    }
    int pop_front(){
        if(data[0]==data_max[0])
            data_max.pop_front();
        int temp = data.front();
        data.pop_front();
        return temp;
    }
};

//int main(){
//    QueueMax q;
//
////    for (int i = 0; i < n1.size(); ++i) {
////        cout<< n1[i];
////    }
//    system("PAUSE");
//    return 0;
//}
